### PR TITLE
Here it is: the PR to merge the error handling staging branch into develop

### PR DESCRIPTION
**Pull request by balos1**
🙌

### CODE REVIEW

**Comment by balos1** [comment]
There are 6 commits at the very end of this PR (after the #391 merge commit) that fixed bugs that prevented the CI from passing. You can click on each of those commits to see those additional changes.

File: cmake/SundialsSetupCompilers.cmake [hunk 1]
```diff
@@ -476,20 +476,42 @@ endforeach()
...
 
+set(SUNDIALS_DEV_CLANG_TIDY_DIR ${CMAKE_BINARY_DIR}/clang-tidy/)
+
 if(SUNDIALS_DEV_CLANG_TIDY)
-  make_directory(${CMAKE_BINARY_DIR}/clang-tidy)
+  find_program(CLANG_TIDY_PATH NAMES clang-tidy)
+  if(NOT CLANG_TIDY_PATH)
+      message(FATAL_ERROR "Could not find the program clang-tidy")
+  endif()
+  message("Found IWYU: ${CLANG_TIDY_PATH}")
```
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:486]:
```suggestion
  message("Found clang-tidy: ${CLANG_TIDY_PATH}")
```

File: cmake/SundialsSetupCompilers.cmake [hunk 2]
```diff
@@ -476,20 +476,42 @@ endforeach()
...
+  make_directory(${SUNDIALS_DEV_IWYU_DIR})
```
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:487]: Move below of IWYU block below
```suggestion
```

File: cmake/SundialsSetupCompilers.cmake [hunk 3]
```diff
@@ -476,20 +476,42 @@ endforeach()
...
   endif()
 endif()
+
+if(SUNDIALS_DEV_IWYU)
+  find_program(IWYU_PATH NAMES include-what-you-use iwyu)
+  if(NOT IWYU_PATH)
+    message(FATAL_ERROR "Could not find the program include-what-you-use")
+  endif()
+  message("Found IWYU: ${IWYU_PATH}")
+  set(CMAKE_C_INCLUDE_WHAT_YOU_USE ${IWYU_PATH}
```
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:511]:
```suggestion
  make_directory(${SUNDIALS_DEV_IWYU_DIR}) 
  set(CMAKE_C_INCLUDE_WHAT_YOU_USE ${IWYU_PATH}
```

File: cmake/SundialsSetupCompilers.cmake [hunk 4]
```diff
@@ -476,20 +476,42 @@ endforeach()
...
 if(SUNDIALS_DEV_CLANG_TIDY)
-  make_directory(${CMAKE_BINARY_DIR}/clang-tidy)
+  find_program(CLANG_TIDY_PATH NAMES clang-tidy)
+  if(NOT CLANG_TIDY_PATH)
+      message(FATAL_ERROR "Could not find the program clang-tidy")
+  endif()
+  message("Found IWYU: ${CLANG_TIDY_PATH}")
+  make_directory(${SUNDIALS_DEV_IWYU_DIR})
+  make_directory(${SUNDIALS_DEV_CLANG_TIDY_DIR})
   if(SUNDIALS_DEV_CLANG_TIDY_FIX_ERRORS)
```
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:489]: Should this be added as an advanced option?

File: scripts/iwyu.imp [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    include: ["@<sundials/sundials_.*hpp>", "private", "<sundials/sundials_core.hpp>", "public"],
+    include: ["@\"sundials/sundials_.*hpp\"", "private", "<sundials/sundials_core.hpp>", "public"],
+    include: ["<sundials/priv/sundials_context_impl.h>", "private", "<sundials/sundials_core.h>", "public"]
+    include: ["<sundials/priv/sundials_errors_impl.h>", "private", "<sundials/sundials_core.h>", "public"]
+    include: ["<sundials/priv/sundials_mpi_errors_impl.h>", "private", "<sundials/sundials_core.h>", "public"]
+    include: ["\"sundials/priv/sundials_context_impl.h\"", "private", "<sundials/sundials_core.h>", "public"]
+    include: ["\"sundials/priv/sundials_errors_impl.h\"", "private", "<sundials/sundials_core.h>", "public"]
+    include: ["\"sundials/priv/sundials_mpi_errors_impl.h\"", "private", "<sundials/sundials_core.h>", "public"]
+  }
+]
```
Comment (gardner48) [inline scripts/iwyu.imp:14]: Add new line at end of file

**Review by gardner48 — CHANGES_REQUESTED** [review]

File: cmake/SundialsSetupCompilers.cmake (hunk 3 above)
Comment (gardner48) [reply cmake/SundialsSetupCompilers.cmake:511]: The variable `SUNDIALS_DEV_IWYU_DIR` is unset

File: cmake/SundialsSetupCompilers.cmake [hunk 6]
```diff
@@ -476,20 +476,42 @@ endforeach()
...
-      --export-fixes=${CMAKE_B_DIR}/clang-tidy/clang-tidy-cxx-fixes.yaml
+    set(CMAKE_CXX_CLANG_TIDY
+      ${CLANG_TIDY_PATH}
+      -format-style='file'
+      --export-fixes=${SUNDIALS_DEV_CLANG_TIDY_DIR}/clang-tidy-cxx-fixes.yaml
     )
   endif()
 endif()
+
+if(SUNDIALS_DEV_IWYU)
```
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:505]:
```suggestion
set(SUNDIALS_DEV_IWYU_DIR ${CMAKE_BINARY_DIR}/iwyu/)

if(SUNDIALS_DEV_IWYU)
```

File: cmake/SundialsSetupCompilers.cmake [hunk 7]
```diff
@@ -476,20 +476,42 @@ endforeach()
...
+  find_program(IWYU_PATH NAMES include-what-you-use iwyu)
+  if(NOT IWYU_PATH)
+    message(FATAL_ERROR "Could not find the program include-what-you-use")
+  endif()
+  message("Found IWYU: ${IWYU_PATH}")
```
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:510]:
```suggestion
  message(STATUS "Found IWYU: ${IWYU_PATH}")
```

File: cmake/SundialsSetupCompilers.cmake (hunk 6 above)
Comment (gardner48) [reply cmake/SundialsSetupCompilers.cmake:505]: Actually, is `SUNDIALS_DEV_IWYU_DIR` needed? IWYU doesn't seem to have an option to write its suggestions to a file.

File: cmake/SundialsSetupCompilers.cmake (hunk 3 above)
Comment (gardner48) [reply cmake/SundialsSetupCompilers.cmake:511]: Maybe this isn't needed, see comment below.

**Review by drreynolds — DISMISSED** [review]
I'll defer to @gardner48 for the CMake, GitHub actions, "iwyu", tidy and format changes.  From what else I see, this looks great.

File: cmake/SundialsSetupCompilers.cmake (hunk 4 above)
Comment (balos1) [reply cmake/SundialsSetupCompilers.cmake:483]: There is no reason it should need to be changed, so I think its best to leave it out for now to reduce the number of options that display for developers.

File: cmake/SundialsSetupCompilers.cmake (hunk 6 above)
Comment (balos1) [reply cmake/SundialsSetupCompilers.cmake:499]: As you have noted, it is not needed.

File: cmake/SundialsSetupCompilers.cmake [hunk 8]
```diff
@@ -111,19 +111,19 @@ endif()
...
 message(STATUS "C extensions set to ${CMAKE_C_EXTENSIONS}")
 
-# Profiling generally requires ISO C99 or newer for __func__ though some
-# compilers define __func__ even with ISO C90.
-if(SUNDIALS_BUILD_WITH_PROFILING AND (CMAKE_C_STANDARD STREQUAL "90"))
-  message(WARNING "SUNDIALS_BUILD_WITH_PROFILING=ON requires __func__, compilation may fail with CMAKE_C_STANDARD=90")
+# Before we do compile checks, ensure warnings are reported as errors
+set(OLD_CMAKE_REQUIRED_FLAGS ${CMAKE_REQUIRED_FLAGS})
+if(NOT ENABLE_WARNINGS_AS_ERRORS)
+  set(CMAKE_REQUIRED_FLAGS "-Werror")
```
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:126]: Will this flag cause any problems on Windows or compilers other than gcc/clang?

**Review by gardner48 — COMMENTED** [review]
One question about the warning flag, otherwise this looks good to me

**Review by gardner48 — APPROVED** [review]

File: cmake/SundialsSetupCompilers.cmake (hunk 8 above)
Comment (balos1) [reply cmake/SundialsSetupCompilers.cmake]: As per our offline discussions, this would cause a problem so we removed it.

**Pull request merged**